flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceStreamTask.java

```diff
@@ -118,7 +124,9 @@ protected void performDefaultAction(ActionContext context) throws Exception {
 		}
 
 		sourceThread.join();
-		sourceThread.checkThrowSourceExecutionException();
+		if (!isFinished) {
```

Comment:
Given the above comment, this can be a `protected` getter method for the `finished` flag in the `StreamTask`, available to all `StreamTask`s.